File: Semester1/Modules/stack/stack.cpp
#include <iostream>
#include "stack.h"

using namespace std;

struct StackElement
{
    char value;
    StackElement *next;
};

struct Stack
{
    StackElement *head;
};

Stack *createStack()
{
     Stack *stack = nullptr;
     stack = new Stack;
     stack->head = nullptr;
     return stack;
}

StackElement *createStackElement(char value)
{
    StackElement *stackElement = new StackElement;
    stackElement->next = nullptr;
    stackElement->value = value;
    return stackElement;
}

void push(char value, Stack *stack)
{
    if (stack->head == nullptr)
        stack->head = createStackElement(value);
    else
    {
        StackElement *auxilary = nullptr;
        auxilary = createStackElement(value);
        auxilary->next = stack->head;
        stack->head = auxilary;
    }
}

bool isEmpty(Stack *stack)
{
    return (stack->head == nullptr);
}

char pop(Stack *stack)
{
    char value = stack->head->value;
    StackElement *auxilary = stack->head;
    stack->head = stack->head->next;
    delete auxilary;
    return value;
}


void deleteStack(Stack *stack)
{
    while(stack->head != nullptr)
    {
        StackElement *auxilary = stack->head;
        stack->head = stack->head->next;
        delete auxilary;
    }
    delete stack;
}

void printStack(StackElement *stackElement)
{
    if (stackElement == nullptr)
        return;

    printStack(stackElement->next);
    cout << stackElement->value << " ";
}


void printStack(Stack *stack)
{
    if (stack->head == nullptr)
        return;
    else
    {
        printStack(stack->head);
        cout << endl;
    }
}

File: Semester2/2/task1/task1/Stack.cs
using System;

namespace Task1
{
    class Stack
    {

        private StackElement head;

        public void Push(int value)
        {
            StackElement newElement = new StackElement(value, head);
            head = newElement;
        }

        public int Pop()
        {
            int value = head.Value;
            head = head.Next;
            return value;
        }

        public bool IsEmpty()
        {
            return (head == null);
        }

        public void PrintStack()
        {
            StackElement auxilary = head;
            while (auxilary != null)
            {
                Console.Write(auxilary.Value + " ");
                auxilary = auxilary.Next;
            }
            Console.WriteLine();
        }

        private class StackElement
        {
            public StackElement(int value, StackElement next)
            {
                this.Value = value;
                this.Next = next;
            }

            public int Value { get; set; }
            public StackElement Next { get; set; }
        }
    }
}
